### PR TITLE
Fix Quality Report reproducibility 

### PR DESCRIPTION
**Pull request by zurk**
Based on PR https://github.com/src-d/style-analyzer/pull/517.
Closes https://github.com/src-d/style-analyzer/issues/557.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -61,7 +63,11 @@ report-noisy: $(NOISY_REPORT_DIR)
 		2>&1 | tee $(NOISY_REPORT_DIR)/logs.txt
 report-quality: $(QUALITY_REPORT_DIR)
 	python3 -m lookout.style.format.benchmarks.top_repos_quality -o $(QUALITY_REPORT_DIR) \
-		2>&1 | tee $(QUALITY_REPORT_DIR)/logs.txt
+		-i $(QUALITY_REPORT_REPOS) 2>&1 | tee $(QUALITY_REPORT_DIR)/logs.txt
 report-release: report-smoke report-noisy report-quality
 
-.PHONY: check docker-test bblfsh-start report-smoke report-noisy report-quality report-release
+expected-vnode-number:
```
Comment (vmarkovtsev) [inline Makefile]:
```suggestion
expected-vnodes-number:
```

File: lookout/style/format/benchmarks/data/quality_report_repos.csv [hunk 2]
```diff
@@ -0,0 +1,20 @@
+url,to,from
+https://github.com/30-seconds/30-seconds-of-code,3a122c9cfcbdc091227879a06a32bc67ccd0d35d,c8c60895e80b8bc90583502accdaa339b794609c
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/data/quality_report_repos.csv]: Can you please sort this list by size, smallest first?

File: lookout/style/format/benchmarks/expected_vnode_number.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            "select_features_number": 250,
+            "return_sibling_indices": False,
+        },
+        "trainable_rules": {
+            "prune_attributes": False,
+            "uncertain_attributes": False,
+            "n_estimators": 3,
+        },
+        "optimizer": {
+            "n_iter": 6,
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnode_number.py]: It is always >=10, so you can set it to 1 with the same success

File: lookout/style/format/benchmarks/expected_vnode_number.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        "feature_extractor": {
+            "left_siblings_window": 2,
+            "right_siblings_window": 2,
+            "parents_depth": 2,
+            "select_features_number": 250,
+            "return_sibling_indices": False,
+        },
+        "trainable_rules": {
+            "prune_attributes": False,
+            "uncertain_attributes": False,
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnode_number.py]: We no longer have uncertain_attributes

File: lookout/style/format/benchmarks/expected_vnode_number.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            git_dir = ensure_repo(repository, tmpdirname)
+            server.run("push", fr=from_commit, to=to_commit, port=port, git_dir=git_dir,
+                       log_level="warning", bblfsh=bblfsh,
+                       config_json=json.dumps(train_config))
+    finally:
+        FeatureExtractor._log.debug = FeatureExtractor._log.debug.original
+    return expected_vnodes_number
+
+
+def calc_expected_vnode_number_entry(input: str, output: str, runs: int, bblfsh: Optional[str],
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnode_number.py]:
```suggestion
def calc_expected_vnodes_number_entry(input: str, output: str, runs: int, bblfsh: Optional[str],
```

File: lookout/style/format/benchmarks/expected_vnode_number.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                                        "of %d. Set to nan.", cur_run + 1, runs, vnodes_number,
+                                        row["vnodes_number"])
+                            row["vnodes_number"] = float("nan")
+                        else:
+                            row["vnodes_number"] = vnodes_number
+                    except Exception:
+                        log.exception("-" * 20 + "\nFailed to process %s repo", row["url"])
+                        continue
+        if cur_run != runs - 1:
+            log.info("Give bblfsh %d sec to rest...", time_to_sleep)
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnode_number.py]: I am afraid the resting will not help here... We need to ask the user to restart I think

File: lookout/style/format/benchmarks/top_repos_quality.py [hunk 7]
```diff
@@ -277,6 +257,26 @@ def _get_json_data(report: str) -> dict:
     return data
 
 
+def handle_input_arg(input_arg: str,
+                     log: Optional[logging.Logger] = None) -> str:
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/top_repos_quality.py]:
```suggestion
                     log: Optional[logging.Logger] = None) -> Iterator[str]:
```

File: lookout/style/format/cmdline_tools.py [hunk 8]
```diff
@@ -157,6 +159,25 @@ def add_parser(name, help):
...
+    calc_expected_vnode.add_argument(
+        "-i", "--input", required=True,
+        help="csv file with repositories for quality report."
+             "Should contain url, to and from columns.")
+    calc_expected_vnode.add_argument(
+        "-o", "--output", required=True, help="Path to a output csv file.")
+    calc_expected_vnode.add_argument(
+        "-b", "--bblfsh", help="Bblfsh address to use.")
+    calc_expected_vnode.add_argument(
+        "-r", "--runs", default=3, help="Repeat number to ensure the result correctness.")
```
Comment (vmarkovtsev) [inline lookout/style/format/cmdline_tools.py]:
```suggestion
        "-r", "--runs", default=3, help="Number of repeats to ensure the result's correctness.")
```

**Comment by zurk** [comment]
🎉 
ready for review

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
+"""Util to workaround https://github.com/src-d/style-analyzer/issues/557."""
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]:
```suggestion
"""Utility to workaround https://github.com/src-d/style-analyzer/issues/557."""
```

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+analyzer_class = FormatAnalyzer
+
+
+def _stop_bblfshd():
+    subprocess.check_call("docker rm -f expected_vnodes_bblfshd".split(),
+                          stdout=sys.stderr, stderr=sys.stderr)
+
+
+def _restart_bblfshd(first: bool=False) -> str:
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]: `first` what? (rename)

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    log = logging.getLogger("expected_vnodes_number")
+    port = server.find_port()
+    log.info("Restarting bblfshd")
+    if not first:
+        _stop_bblfshd()
+    subprocess.check_call(("docker run -d --rm --name expected_vnodes_bblfshd --privileged -p "
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]: We already have `docker` Python package in our deps. It can do the same - can you please change subprocess calls to API calls

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+                          stdout=subprocess.DEVNULL, stderr=sys.stderr)
+    address = "localhost:%d" % port
+    log.info("bblfshd available at %s" % address)
+    return address
+
+
+def get_vnodes_number(repository: str, from_commit: str, to_commit: str, port: int,
+                      bblfsh: Optional[str]) -> int:
+    """
+    Generate `QualityReport` for a repository. If it fails it returns empty reports.
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]: This docstring seems irrelevant to what the function does...

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+    :param repository: URL of repository.
+    :param from_commit: Hash of commit.
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]:
```suggestion
    :param from_commit: Hash of the base commit.
```

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    :param to_commit: Hash of commit.
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]:
```suggestion
    :param to_commit: Hash of the head commit.
```

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    :param port: Port for QualityReportAnalyzer.
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]: For QualityReportAnalyzer? 🤔 Do you mean for Babelfish server which is used by QualityReportAnalyzer?

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+    :param bblfsh: Babelfish server address to use. Specify None to use the default value.
+    :return: Reports.
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]: Nope, we return an integer

File: lookout/style/format/benchmarks/expected_vnodes_number.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+    def capture_debug_msg(log_debug):
+        @functools.wraps(log_debug)
+        def wrapped_log(msg, *args, **kwargs):
+            if msg != FeatureExtractor.PARSED_VNODES_MSG:
+                return log_debug(msg, *args, **kwargs)
+            nonlocal expected_vnodes_number
+            if expected_vnodes_number != -1:
+                raise RuntimeError("%s should be called only one time." % log_debug.__name__)
+            log_debug(msg, *args, **kwargs)
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/expected_vnodes_number.py]: Wow damn this is too hacky. Please solve it better. We run everything in the same process, so there should not be huge difficulties. Examples:

- Override a method
- Add a callback
- Subclass and override

File: lookout/style/format/cmdline_tools.py [hunk 18]
```diff
@@ -173,6 +175,24 @@ def add_parser(name, help):
     rule_parser.add_argument("model", help="Path to the model file.")
     rule_parser.add_argument("hash", help="Hash of the rule (8 chars).")
 
+    # Should be removed as soon as https://github.com/src-d/style-analyzer/issues/557 resolved.
```
Comment (vmarkovtsev) [inline lookout/style/format/cmdline_tools.py]:
```python
# FIXME(zurk): remove when https://github.com/src-d/style-analyzer/issues/557 is resolved
```

File: lookout/style/format/cmdline_tools.py [hunk 19]
```diff
@@ -173,6 +175,24 @@ def add_parser(name, help):
...
+    calc_expected_vnodes = add_parser("calc-expected-vnodes-number",
+                                      "Get a csv file with an expected number of nodes set. "
```
Comment (vmarkovtsev) [inline lookout/style/format/cmdline_tools.py]: Write the CSV file with expected numbers of virtual nodes extracted from repositories.

File: lookout/style/format/cmdline_tools.py [hunk 20]
```diff
@@ -173,6 +175,24 @@ def add_parser(name, help):
...
+                                      "Required for quality report generation. It is a workaround "
+                                      "for https://github.com/src-d/style-analyzer/issues/557. "
+                                      "Docker is required.")
```
Comment (vmarkovtsev) [inline lookout/style/format/cmdline_tools.py]:
```suggestion
                                      "Docker service is required to be running.")
```

File: lookout/style/format/cmdline_tools.py [hunk 21]
```diff
@@ -173,6 +175,24 @@ def add_parser(name, help):
...
+    calc_expected_vnodes.set_defaults(handler=calc_expected_vnodes_number_entry)
+    calc_expected_vnodes.add_argument(
+        "-i", "--input", required=True,
+        help="csv file with repositories for quality report."
```
Comment (vmarkovtsev) [inline lookout/style/format/cmdline_tools.py]:
```suggestion
        help="CSV file with repositories for quality report."
```

File: lookout/style/format/benchmarks/expected_vnodes_number.py (hunk 15 above)
Comment (zurk) [reply lookout/style/format/benchmarks/expected_vnodes_number.py]: No, but there is a mistake anyway. Should be `FormatAnalyzer`. Thanks!

File: lookout/style/format/benchmarks/expected_vnodes_number.py (hunk 16 above)
Comment (zurk) [reply lookout/style/format/benchmarks/expected_vnodes_number.py]: yeah, I am a victim of a copy-paste.

File: lookout/style/format/feature_extractor.py [hunk 22]
```diff
@@ -264,6 +267,8 @@ def extract_features(self, files: Iterable[File], lines: List[List[int]]=None) \
 
         vnodes_parsed_number = sum(len(vn) for vn, _, _ in parsed_files)
         self._log.debug("Parsed %d vnodes", vnodes_parsed_number)
+        if self.vnodes_number_callback is not None:
+            self.vnodes_number_callback(vnodes_parsed_number)
```
Comment (vmarkovtsev) [inline lookout/style/format/feature_extractor.py]: I am going to split `extract_features` into several functions and you will not have to add this callback. Stay tuned.

**Comment by vmarkovtsev** [comment]
@zurk Once we merge https://github.com/src-d/style-analyzer/pull/589 you will be able to

```python
original_files_to_xy = FeatureExtractor._files_to_xy

def hacked_files_to_xy(self, parsed_files):
    vnodes_parsed_number = sum(len(vn) for vn, _, _ in parsed_files)
    # do whatever with it
    return original_files_to_xy(self, parsed_files)

FeatureExtractor._files_to_xy = hacked_files_to_xy
```

**Comment by zurk** [comment]
Good!

**Comment by zurk** [comment]
@vmarkovtsev done.
